pkg/controller/restore_controller.go

```diff
@@ -462,43 +468,118 @@ func mostRecentCompletedBackup(backups []*api.Backup) *api.Backup {
+		return backupInfo{}, errors.Wrap(err, "error initializing object store")
+	}
+	info.bucketName = location.Spec.ObjectStorage.Bucket
 
-	logContext.Debug("Backup not found in backupLister, checking object storage directly")
-	backup, err = c.getBackup(objectStore, c.bucket, name)
+	return info, nil
+}
+
+// fetchFromBackupStorage checks each backup storage location, starting with the default,
+// looking for a backup that matches the given backup name.
+func (c *restoreController) fetchFromBackupStorage(backupName string, pluginManager plugin.Manager) (backupInfo, error) {
+	locations, err := c.backupLocationLister.BackupStorageLocations(c.namespace).List(labels.Everything())
 	if err != nil {
-		return nil, err
+		return backupInfo{}, err
+	}
+
+	orderedLocations := orderedBackupLocations(locations, c.defaultBackupLocation)
+
+	logContext := c.logger.WithField("backupName", backupName)
+	var info backupInfo
+	for _, location := range orderedLocations {
+		info, err = c.backupInfoForLocation(location, backupName, pluginManager)
+		if err == nil {
+			break
+		}
+		log := logContext.WithField("locationName", location.Name)
+		log.WithError(errors.WithStack(err)).Error("Unable to fetch a backup from object storage")
+	}
+
+	if info == (backupInfo{}) {
+		return backupInfo{}, errors.New("not able to fetch from backup storage")
+	}
+
+	return info, nil
+}
+
+func orderedBackupLocations(locations []*api.BackupStorageLocation, defaultLocationName string) []*api.BackupStorageLocation {
+	var result []*api.BackupStorageLocation
+
+	for i := range locations {
+		if locations[i].Name == defaultLocationName {
+			// put the default location first
+			result = append(result, locations[i])
+			// append everything before the default
+			result = append(result, locations[:i]...)
+			// append everything after the default
+			result = append(result, locations[i+1:]...)
+
+			return result
+		}
+	}
+
+	return locations
+}
+
+func (c *restoreController) backupInfoForLocation(location *api.BackupStorageLocation, backupName string, pluginManager plugin.Manager) (backupInfo, error) {
+	objectStore, err := getObjectStoreForLocation(location, pluginManager)
+	if err != nil {
+		return backupInfo{}, err
+	}
+
+	backup, err := c.getBackup(objectStore, location.Spec.ObjectStorage.Bucket, backupName)
+	if err != nil {
+		return backupInfo{}, err
 	}
 
 	// ResourceVersion needs to be cleared in order to create the object in the API
 	backup.ResourceVersion = ""
-	// Clear out the namespace too, just in case
+	// Clear out the namespace, in case the backup was made in a different cluster, with a different namespace
 	backup.Namespace = ""
 
-	created, createErr := c.backupClient.Backups(c.namespace).Create(backup)
-	if createErr != nil {
-		logContext.WithError(errors.WithStack(createErr)).Error("Unable to create API object for Backup")
-	} else {
-		backup = created
+	backupCreated, _ := c.backupClient.Backups(c.namespace).Create(backup)
+	if err == nil {
```

Comment:
I really like the code like that. But question about logic, as per our other thread: if the create fails, should we not return the backup we had fetched anyway? If the answer is yes, then the above would fail this logic because in the case of a failure on `create` `backupCreated` would not be a clean backup. Right?

Reply:
If the create fails, we should return the `backupInfo` zero value + the error.

Reply:
Re-read your comments about this. I took `we've already found the backup we're looking for. ` to mean, "we're good with the found backup", when in reality it means we found it but it didn't work, so just ditch it. Copy that.